core/rpc/src/state/mod.rs

```diff
@@ -258,25 +258,30 @@ impl<B, E, Block: BlockT, RA> State<B, E, Block, RA> where
 		&self,
 		range: &QueryStorageRange<Block>,
 		keys: &[StorageKey],
+		last_values: &mut HashMap<StorageKey, Option<StorageData>>,
 		changes: &mut Vec<StorageChangeSet<Block::Hash>>,
 	) -> Result<()> {
-		let mut last_state: HashMap<_, Option<_>> = Default::default();
 		for block in range.unfiltered_range.start..range.unfiltered_range.end {
 			let block_hash = range.hashes[block].clone();
 			let mut block_changes = StorageChangeSet { block: block_hash.clone(), changes: Vec::new() };
 			let id = BlockId::hash(block_hash);
 			for key in keys {
 				let (has_changed, data) = {
 					let curr_data = self.client.storage(&id, key)?;
-					let prev_data = last_state.get(key).and_then(|x| x.as_ref());
+					let prev_data = last_values.get(key).and_then(|x| x.as_ref());
 					(curr_data.as_ref() != prev_data, curr_data)
 				};
 				if has_changed {
 					block_changes.changes.push((key.clone(), data.clone()));
 				}
-				last_state.insert(key.clone(), data);
+				last_values.insert(key.clone(), data);
+			}
+
+			// always insert entry for the first block, because we're claiming to provide values of
```

Comment:
The RPC method description claims that "This first returned result contains the initial state of storage for all keys.". But actually the first entry won't have a value for the key if the value is `None`:
```json
{"jsonrpc":"2.0","result":[{"block":"0x61..0c","changes":[]}],"id":1}
```
But if later the key has been changed to `Some(_)` and then back to `None`, the changes will look like that (mind `null` in the last block' changes):
```json
{"jsonrpc":"2.0","result":[{"block":"0x61..0c","changes":[]},{"block":"0x81..8d","changes":[["0x05","0x06"]]},{"block":"0x91..9e","changes":[["0x05",null]]}],"id":1}
```
So we are using both absence-of-key and `null` to mark `None` values. What do you think - does it worth to use only `null` for that? If so, I could revert last commit, that itself reverts this fix.

Reply:
It might be a bit confusing but it seems fine for me. The alternative would be to print out all the keys with `null` value in the initial block, right?
While we could do that if user subscribes to specific keys (still a bit wasteful), it would not be feasible in case of a wildcard.

I'm fine with both, maybe let's hear @jacogr's opinion on this.

Reply:
> The alternative would be to print out all the keys with null value in the initial block, right?

Exactly

Reply:
Prefer consistency, so always showing "something" - currently on the middleware side we have to handle both `undefined` and `null` - either way, as long as the order is correct, _should_ not break anything.

Reply:
Thanks, so it'll be `null` everywhere (speaking of this method only, of course)